Source/WebCore/platform/network/curl/OpenSSLHelper.cpp

```diff
@@ -301,19 +301,64 @@ static void getSubjectAltName(const X509* x509, Vector<String>& dnsNames, Vector
             if (value->d.iPAddress->length == 4)
                 ipAddresses.append(makeString(data[0], ".", data[1], ".", data[2], ".", data[3]));
             else if (value->d.iPAddress->length == 16) {
-                StringBuilder ipAddress;
-                for (int i = 0; i < 8; i++) {
-                    ipAddress.append(makeString(hex(data[0] << 8 | data[1], 4)));
-                    if (i != 7)
-                        ipAddress.append(":");
-                    data += 2;
-                }
-                ipAddresses.append(ipAddress.toString());
+                Span<uint8_t, 16> dataSpan { data, 16 };
+                ipAddresses.append(canonicalizeIPv6Address(dataSpan));
             }
         }
     }
 }
 
+String canonicalizeIPv6Address(Span<uint8_t, 16> data)
+{
+    bool compressCurrentSection = false;
+    size_t maxZeros = 0;
+    std::optional<size_t> startRunner;
+    std::optional<size_t> endRunner;
+    std::optional<size_t> start;
+    std::optional<size_t> end;
+
+    for (int i = 0; i < 8; i++) {
+        compressCurrentSection = !data[2 * i] && !data[2 * i + 1];
+        if (compressCurrentSection) {
+            startRunner = !startRunner.has_value() ? i : startRunner;
+            endRunner = i;
+            size_t len = endRunner.value() - startRunner.value() + 1;
+            if (len > maxZeros) {
+                start = startRunner;
+                end = endRunner;
+                maxZeros = len;
+            }
+        } else
+            startRunner.reset();
+    }
+
+    size_t minimum = 1;
+    StringBuilder ipAddress;
+    String oldSection = ""_s;
+    StringBuilder newSection;
+    for (int j = 0; j < 8; j++) {
+        if (j == start && maxZeros > minimum) {
+            if (ipAddress.isEmpty())
+                ipAddress.append(":");
+            ipAddress.append(":");
+
+            j = end.value();
+            continue;
+        }
+        oldSection = makeString(hex(data[2 * j] << 8 | data[2 * j + 1], 4, Lowercase));
```

Comment:
This still does several unnecessary string allocations.  This could be an integer operation.
Much better than it was before, and since it's only used in a rare case in the cURL code, I think it's fine to add it.